File: haskell/MK.hs
-- | A Haskell version of mk.rkt.
--
-- The structure follows the Racket source section by section.  Three
-- places had to be adapted rather than transliterated:
--
--   * Racket distinguishes fresh logic variables by @eq?@ (struct
--     identity).  Haskell has no such notion, so a variable carries an
--     integer and goals thread a counter alongside the substitution.
--
--   * @delay@/@force@ are kept explicit even though Haskell is lazy: the
--     interleaving in 'mplus' works by /observing/ that a stream is
--     suspended, which a transparent thunk would not allow.
--
--   * @name-prefix@ was a mutable global set by @set-name-prefix@; here
--     it is a parameter of 'reifyWith' ('reify' uses the default \"_\").
--     @display-code@/@debug-display@ printed the source of a @run@ form,
--     which is not available without macros, and is dropped.
module MK where

import qualified Debug.Trace as Debug


--------------------- substitution ----------------------

newtype Var = Var Int
  deriving (Eq)

data Term
  = TVar Var
  | TSym String
  | TStr String
  | TNum Integer
  | TBool Bool
  | TNil
  | TPair Term Term
  deriving (Eq)

type S = [(Var, Term)]

emptyS :: S
emptyS = []

sizeS :: S -> Int
sizeS = length

extS :: Var -> Term -> S -> S
extS x v s = (x, v) : s

walk :: Term -> S -> Term
walk v s =
  case v of
    TVar x ->
      case lookup x s of
        Nothing -> v
        Just p  -> walk p s
    _ -> v

walkStar :: Term -> S -> Term
walkStar v s =
  case walk v s of
    TVar x      -> TVar x
    TPair a d   -> TPair (walkStar a s) (walkStar d s)
    v'          -> v'

unify :: Term -> Term -> S -> Maybe S
unify u v s =
  case (walk u s, walk v s) of
    (TVar x, TVar y) | x == y -> Just s
    (TVar x, v')              -> Just (extS x v' s)
    (u', TVar y)              -> Just (extS y u' s)
    (TPair a d, TPair a' d')  -> unify a a' s >>= unify d d'
    (u', v') | u' == v'       -> Just s
             | otherwise      -> Nothing

namePrefix :: String
namePrefix = "_"

name :: String -> Int -> Term
name prefix n = TSym (prefix ++ show n)

reifyS :: String -> Term -> S -> S
reifyS prefix v s =
  case walk v s of
    TVar x    -> extS x (name prefix (sizeS s)) s
    TPair a d -> reifyS prefix d (reifyS prefix a s)
    _         -> s

reifyWith :: String -> Term -> S -> Term
reifyWith prefix v s =
  let v' = walkStar v s
  in walkStar v' (reifyS prefix v' emptyS)

reify :: Term -> S -> Term
reify = reifyWith namePrefix


---------------------- composition ----------------------

-- | A goal runs against a substitution paired with the counter that
-- supplies fresh variables.
data State = State { subst :: S, counter :: Int }

-- | A goal maps a state to a stream of states: 'Stream' instantiated at
-- the substitution it threads.
type Goal = State -> Stream State

-- | The stream is parameterised by what it carries.  @Delay@ is
-- @thunk@: its field is lazy, so the suspension is real, but the
-- constructor keeps it visible to 'mplus' and 'bind'.
data Stream a
  = Empty
  | Delay (Stream a)
  | Cons a (Stream a)

force :: Stream a -> Stream a
force (Delay s) = s
force s         = s

bind :: (a -> Stream b) -> Stream a -> Stream b
bind g v =
  case v of
    Empty      -> Empty
    Delay _    -> Delay (bind g (force v))
    Cons hd tl -> mplus (g hd) (Delay (bind g tl))

-- | @bind*@: the goals are applied left to right.
bindStar :: Stream State -> [Goal] -> Stream State
bindStar v []       = v
bindStar v (g : gs) = bindStar (bind g v) gs

mplus :: Stream a -> Stream a -> Stream a
mplus v f =
  case v of
    Empty      -> f
    Delay _    -> Delay (mplus f (force v))
    Cons hd tl -> Cons hd (Delay (mplus f tl))

mplusStar :: [Stream a] -> Stream a
mplusStar []       = Empty
mplusStar [e]      = e
mplusStar (e : es) = mplus e (Delay (mplusStar es))


------------------- goal constructors -------------------

succeed :: Goal
succeed st = Cons st Empty

-- | @fail@ in mk.rkt; renamed to avoid the clash with 'Prelude.fail'.
failure :: Goal
failure _ = Empty

infix 4 ===

(===) :: Term -> Term -> Goal
(===) u v st =
  Delay
    (case unify u v (subst st) of
       Just s  -> succeed st { subst = s }
       Nothing -> failure st)

-- | Conjunction, the implicit @g0 g ...@ of @exist@ and of a @conde@
-- clause.  Like @bind*@ it inserts no delay of its own.
conj :: [Goal] -> Goal
conj gs st = bindStar (succeed st) gs

-- | The variables of @(exist (x ...) ...)@: 'Term' for one, a tuple of
-- 'Term's for several.
class Fresh a where
  freshVars :: Int -> (a, Int)

instance Fresh Term where
  freshVars n = (TVar (Var n), n + 1)

instance (Fresh a, Fresh b) => Fresh (a, b) where
  freshVars n0 =
    let (a, n1) = freshVars n0
        (b, n2) = freshVars n1
    in ((a, b), n2)

instance (Fresh a, Fresh b, Fresh c) => Fresh (a, b, c) where
  freshVars n0 =
    let (a, n1) = freshVars n0
        (b, n2) = freshVars n1
        (c, n3) = freshVars n2
    in ((a, b, c), n3)

instance (Fresh a, Fresh b, Fresh c, Fresh d) => Fresh (a, b, c, d) where
  freshVars n0 =
    let (a, n1) = freshVars n0
        (b, n2) = freshVars n1
        (c, n3) = freshVars n2
        (d, n4) = freshVars n3
    in ((a, b, c, d), n4)

instance (Fresh a, Fresh b, Fresh c, Fresh d, Fresh e)
      => Fresh (a, b, c, d, e) where
  freshVars n0 =
    let (a, n1) = freshVars n0
        (b, n2) = freshVars n1
        (c, n3) = freshVars n2
        (d, n4) = freshVars n3
        (e, n5) = freshVars n4
    in ((a, b, c, d, e), n5)

-- | @(exist (x ...) g0 g ...)@ is @exist $ \(x, ...) -> [g0, g ...]@.
exist :: Fresh a => (a -> [Goal]) -> Goal
exist f st =
  Delay
    (let (xs, n) = freshVars (counter st)
     in bindStar (succeed st { counter = n }) (f xs))

conde :: [[Goal]] -> Goal
conde cs st =
  Delay (mplusStar [bindStar (succeed st) gs | gs <- cs])

-- | @trace@ in mk.rkt; the variables' source names are not available, so
-- the reified terms are printed under the message alone.
traceG :: String -> [Term] -> Goal
traceG msg vs st =
  Debug.trace
    (unlines [msg ++ show (reify v (subst st)) | v <- vs])
    (succeed st)


----------------------- top level ------------------------

-- | The @take@ of mk.rkt is @take n . streamToList@: laziness makes the
-- bound unnecessary here, and @run*@ needs no infinity.
streamToList :: Stream a -> [a]
streamToList v =
  case v of
    Empty      -> []
    Delay _    -> streamToList (force v)
    Cons hd tl -> hd : streamToList tl

-- | @q@ is bound outside the goal so that the answers can be reified
-- after they are taken from the stream: the stream here is a @Stream
-- State@, so reification happens on the way out rather than inside it.
runAll :: (Term -> [Goal]) -> [Term]
runAll f =
  let q  = TVar (Var 0)
      st = State { subst = emptyS, counter = 1 }
  in map (reify q . subst) (streamToList (Delay (bindStar (succeed st) (f q))))

run :: Int -> (Term -> [Goal]) -> [Term]
run n f = take n (runAll f)


------------------------- terms --------------------------

sym :: String -> Term
sym = TSym

str :: String -> Term
str = TStr

num :: Integer -> Term
num = TNum

bool :: Bool -> Term
bool = TBool

nil :: Term
nil = TNil

infixr 5 `cons`

cons :: Term -> Term -> Term
cons = TPair

list :: [Term] -> Term
list = foldr TPair TNil

instance Show Var where
  show (Var n) = "_." ++ show n

instance Show Term where
  show v =
    case v of
      TVar x       -> show x
      TSym s       -> s
      TStr s       -> show s
      TNum n       -> show n
      TBool True   -> "#t"
      TBool False  -> "#f"
      TNil         -> "()"
      TPair _ _    -> "(" ++ showTail v ++ ")"
    where
      showTail (TPair a d) =
        case d of
          TNil      -> show a
          TPair _ _ -> show a ++ " " ++ showTail d
          _         -> show a ++ " . " ++ show d
      showTail t = show t

File: MK.ts
// A TypeScript version of mk.rkt.
//
// The structure follows the Racket source section by section.  Four
// places had to be adapted rather than transliterated:
//
//   * Racket distinguishes fresh logic variables by `eq?' (struct
//     identity).  JavaScript object identity is the same notion, so
//     `Var' needs no counter: `new Var(...)' is a fresh variable and
//     `===' is `eq?'.
//
//   * `delay'/`force' are ordinary functions rather than a macro, so a
//     suspension is written `delay(() => e)'.  The thunk stays visible
//     in the stream: the interleaving in `mplus' works by *observing*
//     that a stream is suspended.
//
//   * There are no macros, so `exist', `conde' and `run' are functions:
//     the binders of `exist'/`run' are the parameters of a closure and
//     a conjunction is an array of goals.  The source names of those
//     parameters are recovered from the closure for display only; when
//     they cannot be read, positional names are used instead.
//
//   * `display-code'/`debug-display' printed the source of a `run'
//     form, which is not available without macros, and is dropped.
//
// Scheme atoms map to JavaScript as: symbol -> interned `symbol'
// (`sym("a")'), string -> string, number -> number, boolean -> boolean,
// '() -> null.

//--------------------- substitution ----------------------

export class Var {
  readonly name: string;
  constructor(name: string) {
    this.name = name;
  }
}

export class Pair {
  readonly car: Term;
  readonly cdr: Term;
  constructor(car: Term, cdr: Term) {
    this.car = car;
    this.cdr = cdr;
  }
}

export type Atom = symbol | string | number | boolean | null;
export type Term = Var | Atom | Pair;

// An association list of (variable . term), extended at the front.
export interface Frame {
  readonly x: Var;
  readonly v: Term;
  readonly rest: S;
}
export type S = Frame | null;

export const emptyS: S = null;

export function sizeS(s: S): number {
  let n = 0;
  for (let p = s; p !== null; p = p.rest) n++;
  return n;
}

export function extS(x: Var, v: Term, s: S): S {
  return { x, v, rest: s };
}

function assq(x: Var, s: S): Frame | null {
  for (let p = s; p !== null; p = p.rest) {
    if (p.x === x) return p;
  }
  return null;
}

export function walk(v: Term, s: S): Term {
  while (v instanceof Var) {
    const p = assq(v, s);
    if (p === null) return v;
    v = p.v;
  }
  return v;
}

export function walkStar(v: Term, s: S): Term {
  const w = walk(v, s);
  if (w instanceof Pair) {
    return new Pair(walkStar(w.car, s), walkStar(w.cdr, s));
  }
  return w;
}

// `false' rather than `null' is the failure of unification: the empty
// substitution is `null', and both would otherwise be falsy.
export function unify(u: Term, v: Term, s: S): S | false {
  u = walk(u, s);
  v = walk(v, s);
  if (u === v) return s;
  if (u instanceof Var) return extS(u, v, s);
  if (v instanceof Var) return extS(v, u, s);
  if (u instanceof Pair && v instanceof Pair) {
    const s1 = unify(u.car, v.car, s);
    return s1 === false ? false : unify(u.cdr, v.cdr, s1);
  }
  return false;
}

let namePrefix = "_";

export function setNamePrefix(s: string): void {
  namePrefix = s;
}

export function name(n: number): symbol {
  return Symbol.for(namePrefix + n);
}

export function reifyS(v: Term, s: S): S {
  const w = walk(v, s);
  if (w instanceof Var) return extS(w, name(sizeS(s)), s);
  if (w instanceof Pair) return reifyS(w.cdr, reifyS(w.car, s));
  return s;
}

export function reify(v: Term, s: S): Term {
  const w = walkStar(v, s);
  return walkStar(w, reifyS(w, emptyS));
}

//---------------------- composition ----------------------

// The stream is parameterised by what it carries; a goal instantiates
// it at `S', so a goal maps a substitution to a stream of substitutions.
export class SCons<A> {
  readonly head: A;
  readonly tail: Stream<A>;
  constructor(head: A, tail: Stream<A>) {
    this.head = head;
    this.tail = tail;
  }
}

export class Thunk<A> {
  readonly func: () => Stream<A>;
  constructor(func: () => Stream<A>) {
    this.func = func;
  }
}

export type Stream<A> = null | Thunk<A> | SCons<A>;
export type Goal = (s: S) => Stream<S>;

export function delay<A>(f: () => Stream<A>): Stream<A> {
  return new Thunk(f);
}

export function force<A>(th: Stream<A>): Stream<A> {
  return th instanceof Thunk ? th.func() : th;
}

export function bind<A, B>(g: (a: A) => Stream<B>, v: Stream<A>): Stream<B> {
  if (v === null) return null;
  if (v instanceof Thunk) return delay(() => bind(g, force(v)));
  return mplus(g(v.head), delay(() => bind(g, v.tail)));
}

export function bindStar(v: Stream<S>, ...gs: Goal[]): Stream<S> {
  for (const g of gs) v = bind(g, v);
  return v;
}

export function mplus<A>(v: Stream<A>, f: Stream<A>): Stream<A> {
  if (v === null) return f;
  if (v instanceof Thunk) return delay(() => mplus(f, force(v)));
  return new SCons(v.head, delay(() => mplus(f, v.tail)));
}

export function mplusStar<A>(...es: Stream<A>[]): Stream<A> {
  if (es.length === 0) return null;
  const [e, ...rest] = es;
  if (rest.length === 0) return e;
  return mplus(e, delay(() => mplusStar(...rest)));
}

//------------------- goal constructors -------------------

export const succeed: Goal = (s) => new SCons(s, null);

export const fail: Goal = (_s) => null;

// `==' in mk.rkt; `==' is not an operator that can be defined here.
export function eq(u: Term, v: Term): Goal {
  return (s) =>
    delay(() => {
      const s2 = unify(u, v, s);
      return s2 === false ? fail(s) : succeed(s2);
    });
}

// `(exist (x ...) g0 g ...)' is `exist((x, ...) => [g0, g, ...])'.  The
// variables are made inside the suspension, so every entry into the
// goal gets fresh ones.
export function exist(f: (...xs: Term[]) => Goal[]): Goal {
  const names = paramNames(f);
  return (s) =>
    delay(() => {
      const xs = names.map((n) => new Var(n));
      return bindStar(succeed(s), ...f(...xs));
    });
}

export function conde(...cs: Goal[][]): Goal {
  return (s) =>
    delay(() => mplusStar(...cs.map((gs) => bindStar(succeed(s), ...gs))));
}

export function trace(msg: string, ...vs: Term[]): Goal {
  return (s) => {
    for (const v of vs) console.log(`${msg}${show(reify(v, s))}`);
    return succeed(s);
  };
}

//----------------------- top level -----------------------

// Iterative, unlike the Racket `take': JavaScript has no tail calls,
// and forcing a long stream would otherwise grow the stack.
export function take<A>(n: number, v: Stream<A>): A[] {
  const out: A[] = [];
  while (out.length < n) {
    if (v === null) break;
    if (v instanceof Thunk) {
      v = force(v);
      continue;
    }
    out.push(v.head);
    v = v.tail;
  }
  return out;
}

// `q' is bound outside the goal so that the answers can be reified
// after they are taken from the stream: the stream a goal builds is a
// `Stream<S>', so reification happens on the way out rather than in it.
export function run(n: number, f: (q: Term) => Goal[]): Term[] {
  const q = new Var(paramNames(f)[0] ?? "q");
  const v = delay(() => bindStar(succeed(emptyS), ...f(q)));
  return take(n, v).map((s) => reify(q, s));
}

export function runAll(f: (q: Term) => Goal[]): Term[] {
  return run(Infinity, f);
}

//------------------------- terms --------------------------

export function sym(s: string): symbol {
  return Symbol.for(s);
}

export const nil: Term = null;

export function cons(a: Term, d: Term): Pair {
  return new Pair(a, d);
}

export function list(...xs: Term[]): Term {
  return xs.reduceRight<Term>((d, a) => new Pair(a, d), nil);
}

export function show(v: Term): string {
  if (v instanceof Var) return `#(${v.name})`;
  if (v instanceof Pair) return `(${showTail(v)})`;
  if (v === null) return "()";
  switch (typeof v) {
    case "symbol":
      return v.description ?? "";
    case "string":
      return JSON.stringify(v);
    case "boolean":
      return v ? "#t" : "#f";
    default:
      return String(v);
  }
}

function showTail(v: Pair): string {
  const a = show(v.car);
  const d = v.cdr;
  if (d === null) return a;
  if (d instanceof Pair) return `${a} ${showTail(d)}`;
  return `${a} . ${show(d)}`;
}

// The source names of a closure's parameters, for display only.  Any
// shape the reader below does not understand (destructuring, defaults,
// a minified closure) falls back to positional names.
function paramNames(f: Function): string[] {
  const src = f.toString();
  const bare = /^\s*(?:async\s+)?([A-Za-z_$][\w$]*)\s*=>/.exec(src);
  const names = bare
    ? [bare[1]]
    : parenthesised(src)
        .split(",")
        .map((p) => p.trim())
        .filter((p) => p !== "");
  const ok =
    names.length === f.length && names.every((p) => /^[A-Za-z_$][\w$]*$/.test(p));
  return ok ? names : Array.from({ length: f.length }, (_, i) => `x${i}`);
}

function parenthesised(src: string): string {
  const open = src.indexOf("(");
  if (open < 0) return "";
  let depth = 0;
  for (let i = open; i < src.length; i++) {
    if (src[i] === "(") depth++;
    else if (src[i] === ")" && --depth === 0) return src.slice(open + 1, i);
  }
  return "";
}
